### PR TITLE
#199 Upload button: add new component and adapt to use it

### PR DESCRIPTION
**Pull request by 3rdvision**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/ripe-util-vue/issues/199 |
| Decisions | - Split upload-area into [2 components as suggested](https://github.com/ripe-tech/ripe-components-vue/pull/531#issuecomment-914171003) <br> - Add upload-button <br> - Adapt upload-area <br> - Adapt chat and input-image |
| Animated GIF | image |

### CODE REVIEW

File: vue/components/ui/organisms/chat/chat.vue [hunk 1]
```diff
@@ -39,13 +39,17 @@
...
     box-sizing: border-box;
+    display: block;
+    height: initial;
     padding: 20px 20px 20px 20px;
     transition: opacity 0.125s ease-in;
+    white-space: initial;
 }
 
-.chat > .upload-area.dragging {
+.chat > div.upload-area.dragging {
```
Comment (gcandal) [inline vue/components/ui/organisms/chat/chat.vue:52]: Are these changes really needed?

File: vue/components/ui/molecules/upload-button/upload-button.vue [hunk 2]
```diff
@@ -0,0 +1,107 @@
+<template>
+    <div
+        class="upload-button"
```
Comment (gcandal) [inline vue/components/ui/molecules/upload-button/upload-button.vue:3]: Not meant to make you miserable but I believe this is usually the other way around: `button-upload` (e.g.: `input-image`).

File: vue/components/ui/molecules/input-image/input-image.vue [hunk 3]
```diff
@@ -3,57 +3,57 @@
         class="input-image"
         v-bind:multiple="multiple"
         v-bind:accept="accept"
+        ref="uploadArea"
         v-on:update:files="loadFiles"
     >
-        <template v-slot="{ openModal }">
```
Comment (gcandal) [inline vue/components/ui/molecules/input-image/input-image.vue:8]: Can't you keep providing the `openModal` in the slot so that these changes aren't needed?

File: vue/components/ui/molecules/upload-area/upload-area.vue [hunk 4]
```diff
@@ -2,45 +2,37 @@
...
-            </div>
+        <slot>
+            <transition name="fade-in" mode="out-in">
+                <div class="description" v-bind:key="dragging">
+                    {{ descriptionText }}
+                </div>
+            </transition>
+            <icon v-bind:icon="icon" v-bind:size="110" v-bind:color="'gray'" />
         </slot>
+        <upload-button
```
Comment (gcandal) [inline vue/components/ui/molecules/upload-area/upload-area.vue:18]: Couldn't this simply be a substitute of the existing `button-color`, leaving the structure as it was? This change has some issues:
- Being outside the slot means it can't be overridden
- The former point plus the fact that the HTML changed is a breaking change of sorts

File: vue/components/ui/molecules/upload-area/upload-area.vue [hunk 5]
```diff
@@ -121,90 +130,33 @@ export const UploadArea = {
...
-    watch: {
-        files(value) {
-            if (!value || value.length === 0) this.clear();
-            this.filesData = value;
-        },
-        dragging(value) {
-            this.$emit("update:dragging", value);
-        }
+    mounted: function() {
+        this.fileInputRef = this.$refs.uploadButton?.$refs?.filesInput;
```
Comment (gcandal) [inline vue/components/ui/molecules/upload-area/upload-area.vue:142]: Create a `initUploadArea` method in the mixin that does this (setting the `fileInputRef` based on the provided argument), so that you encapsulate the logic there.

File: vue/components/ui/molecules/upload-area/upload-area.vue [hunk 6]
```diff
@@ -121,90 +130,33 @@ export const UploadArea = {
...
-            return base;
         }
     },
     data: function() {
         return {
-            filesData: this.files,
             dragging: false,
-            draggingDisabled: false,
-            dragEnterTarget: null
+            fileInputRef: null
```
Comment (gcandal) [inline vue/components/ui/molecules/upload-area/upload-area.vue:138]: Move to the mixin.

File: vue/components/ui/organisms/chat/chat.vue (hunk 1 above)
Comment (3rdvision) [reply vue/components/ui/organisms/chat/chat.vue:52]: yes, every line

File: vue/components/ui/molecules/upload-area/upload-area.vue (hunk 4 above)
Comment (3rdvision) [reply vue/components/ui/molecules/upload-area/upload-area.vue:18]: The problem is that previously we had this outside the slot:
```
        <input
            type="file"
            hidden
            v-bind:multiple="multiple"
            v-bind:accept="accept"
            ref="filesInput"
            v-on:change="onFilesInputChange"
        />
 ```
 Now that native input is inside `upload-button` we had to move it ouside the slot to replace the native input.
The native input has to somehow be outside the slot for the upload mixin logic to run using it.

An alternative could be moving back `<input>` in this component and toggling it between native or the one from upload-button when we have slots. I didn't do it because it went a little bit against the philosohpy of separating and re-using the components.

Visually, the upload-button becomes invisible with `display: none` when we the component receives slots so there's no visual impact in this change.

**Review by gcandal — APPROVED** [review]

**Comment by joamag** [comment]
I'm merging this one but there's a small quick fix required after this.
In case the `multiple` prop is unset then the default label text should be "Upload File" singular and not plural.
